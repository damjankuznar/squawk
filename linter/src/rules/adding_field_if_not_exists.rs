use crate::{
    versions::Version,
    violations::{RuleViolation, RuleViolationKind},
};

use squawk_parser::ast::{AlterTableCmds, AlterTableType, RawStmt, Stmt};

#[must_use]
pub fn adding_field_if_not_exists(
    tree: &[RawStmt],
    _pg_version: Option<Version>,
    _assume_in_transaction: bool,
) -> Vec<RuleViolation> {
    let mut errs = vec![];

    for raw_stmt in tree {
        match &raw_stmt.stmt {
            Stmt::AlterTableStmt(stmt) => {
                for AlterTableCmds::AlterTableCmd(cmd) in &stmt.cmds {
                    if cmd.subtype == AlterTableType::AddColumn && !cmd.missing_ok {
                        errs.push(RuleViolation::new(
                            RuleViolationKind::AddingFieldIfNotExists,
                            raw_stmt.into(),
                            None,
                        ));
                    }
                }
            }
            _ => continue,
        }
    }
    errs
}

#[cfg(test)]
mod test_rules {
    use crate::{
        check_sql_with_rule,
        versions::Version,
        violations::{RuleViolation, RuleViolationKind},
    };

    use insta::assert_debug_snapshot;

    fn lint_sql(sql: &str, pg_version: Option<Version>) -> Vec<RuleViolation> {
        check_sql_with_rule(
            sql,
            &RuleViolationKind::AddingFieldIfNotExists,
            pg_version,
            false,
        )
        .unwrap()
    }

    #[test]
    fn test_example_bad() {
        let bad_sql = r#"
ALTER TABLE "core_recipe" ADD COLUMN "foo" integer DEFAULT 10;
"#;
        assert_debug_snapshot!(lint_sql(bad_sql, None));
    }

    #[test]
    fn test_example_ok() {
        let ok_sql = r#"
ALTER TABLE "core_recipe" ADD COLUMN IF NOT EXISTS "foo" integer;
        "#;
        assert_debug_snapshot!(lint_sql(ok_sql, None));
    }
}
