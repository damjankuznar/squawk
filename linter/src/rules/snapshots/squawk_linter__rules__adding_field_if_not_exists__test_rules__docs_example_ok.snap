---
source: linter/src/rules/adding_field_if_not_exists.rs
expression: "lint_sql(ok_sql, None)"
---
[]
